File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticContractAttestation.java
package com.foundgine.core.semantic;

import java.util.Locale;
import java.util.Objects;

/** Verifies that a semantic model matches an expected contract fingerprint. */
public final class SemanticContractAttestation {
	private SemanticContractAttestation() {
	}

	public static boolean matches(SemanticModel model, String expectedFingerprint) {
		Objects.requireNonNull(model);
		requireText(expectedFingerprint);
		return model.contractFingerprint().equals(normalize(expectedFingerprint));
	}

	public static void ensureMatches(SemanticModel model, String expectedFingerprint) {
		Objects.requireNonNull(model);
		requireText(expectedFingerprint);
		var expected = normalize(expectedFingerprint);
		if (model.contractFingerprint().equals(expected))
			return;
		throw new IllegalStateException("Semantic contract attestation failed. Expected fingerprint '" + expected
				+ "', but the runtime semantic model has fingerprint '" + model.contractFingerprint() + "'.");
	}

	private static void requireText(String s) {
		if (s == null || s.isBlank())
			throw new IllegalArgumentException("Fingerprint cannot be empty.");
	}

	private static String normalize(String s) {
		var x = s.trim();
		return x.regionMatches(true, 0, "sha256:", 0, 7) ? x.substring(7).toLowerCase(Locale.ROOT)
				: x.toLowerCase(Locale.ROOT);
	}
}

File: src/csharp/Foundgine.Core/Semantic/SemanticContractAttestation.cs
namespace Foundgine.Core.Semantic;

/// <summary>
/// Verifies that a semantic model is the contract expected by a generated or
/// otherwise trusted semantic artifact. The fingerprint is an integrity
/// identifier only; it does not grant authorization.
/// </summary>
public static class SemanticContractAttestation
{
    public static bool Matches(SemanticModel model, string expectedFingerprint)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentException.ThrowIfNullOrWhiteSpace(expectedFingerprint);

        return string.Equals(
            model.ContractFingerprint,
            Normalize(expectedFingerprint),
            StringComparison.Ordinal);
    }

    public static void EnsureMatches(SemanticModel model, string expectedFingerprint)
    {
        ArgumentNullException.ThrowIfNull(model);
        ArgumentException.ThrowIfNullOrWhiteSpace(expectedFingerprint);

        var expected = Normalize(expectedFingerprint);
        if (string.Equals(model.ContractFingerprint, expected, StringComparison.Ordinal))
            return;

        throw new InvalidOperationException(
            $"Semantic contract attestation failed. Expected fingerprint '{expected}', " +
            $"but the runtime semantic model has fingerprint '{model.ContractFingerprint}'.");
    }

    private static string Normalize(string fingerprint)
    {
        const string prefix = "sha256:";
        return fingerprint.StartsWith(prefix, StringComparison.OrdinalIgnoreCase)
            ? fingerprint[prefix.Length..].ToLowerInvariant()
            : fingerprint.ToLowerInvariant();
    }
}